riscv_analysis/src/gen/function_annotations.rs

```diff
@@ -5,13 +5,100 @@ use std::{
 };
 
 use crate::{
-    cfg::{Cfg, Function},
-    parser::Register,
-    parser::{Info, JumpLinkType, LabelString, ParserNode, With},
+    cfg::{CfgNode, Cfg, Function},
+    parser::{Info, JumpLinkType, LabelString, ParserNode, Register, With},
     passes::{CfgError, DiagnosticLocation, GenerationPass},
 };
 
+struct MarkData {
+    pub found: HashSet<Register>,
+    pub instructions: Vec<Rc<CfgNode>>,
+    pub returns: Rc<CfgNode>,
+}
+
 pub struct FunctionMarkupPass;
+
+impl FunctionMarkupPass {
+    fn mark_reachable(entry: Rc<CfgNode>, func: Rc<Function>)
+                      -> Result<MarkData, Box<CfgError>> {
+        // Initialize data for graph search
+        let mut queue = vec![entry.clone()];
+        let mut seen = HashSet::new();      // CFG nodes seen during the traversal
+        let mut defs = HashSet::new();      // Registers this function writes to
+        let mut returns = None;             // Return instructions in this function
+
+        // Traverse the CFG for all nodes reachable from the entry point
+        while let Some(node) = queue.pop() {
+            // Only process each node at most once
+            if seen.contains(&node) {
+                continue;
+            } else {
+                seen.insert(node.clone());
+            }
+
+            // Mark the node as being a part of the given function
+            node.insert_function(func.clone());
+
+            // Collect any registers written to by the node
+            if let Some(dest) = node.node().stores_to() {
+                defs.insert(dest.data);
+            }
+
+            // Collect return instructions
+            if node.node().is_return() {
+                // Set the newly found return to be an jump to the previously
+                // found return.
+                if let Some(ref prev_ret) = returns {
+                    let found_ret = Rc::clone(&node);
+
+                    // Fix the prevs & nexts of both returns
+                    found_ret.clear_nexts();
+                    found_ret.insert_next(Rc::clone(prev_ret));
+                    prev_ret.insert_prev(Rc::clone(&found_ret));
+
+                    // Convert the found return into a jump
+                    let info = Info {
+                        token: crate::parser::Token::Symbol("return".to_string()),
+                        pos: found_ret.node().range().clone(),
+                        file: found_ret.node().file(),
+                    };
+
+                    let inst = With::new(JumpLinkType::Jal, info.clone());
+                    let rd = With::new(Register::X0, info.clone());
+                    let name = With::new(LabelString("__return__".to_string()), info.clone());
+                    let new_node = ParserNode::new_jump_link(
+                        inst,
+                        rd,
+                        name,
+                        prev_ret.node().token(),
+                    );
+                    found_ret.set_node(new_node);
```

Comment:
A little note about this: This is sort of a hack to support multiple return points right now. I should've added a new "function exit" node similar to the "program entry" node and had every return instruction jump to it. 

This doesn't need to be changed for this PR, but I'll add it to the backlog for the future.